helix-core/src/test.rs

```diff
@@ -0,0 +1,161 @@
+                        } else {
+                            left.push(c);
+                        }
+                    }
+                    panic!("missing primary end `]#` {left:?} {s:?}");
+                } else {
+                    while let Some(c) = iter.next() {
+                        if c == '|' {
+                            if let Some(cc) = iter.next_if_eq(&']') {
+                                if iter.next_if_eq(&'#').is_some() {
+                                    primary = Some(ranges.len());
+                                    ranges.push(Range::new(start, left.len()));
+                                    continue 'outer;
+                                } else {
+                                    left.push(c);
+                                    left.push(cc);
+                                }
+                            } else {
+                                left.push(c);
+                            }
+                        } else {
+                            left.push(c);
+                        }
+                    }
+                    panic!("missing primary end `|]#` {left:?} {s:?}");
+                }
+            } else if iter.next_if_eq(&'(').is_some() {
+                if iter.next_if_eq(&'|').is_some() {
+                    while let Some(c) = iter.next() {
+                        if c == ')' && iter.next_if_eq(&'#').is_some() {
+                            ranges.push(Range::new(left.len(), start));
+                            continue 'outer;
+                        } else {
+                            left.push(c);
+                        }
+                    }
+                    panic!("missing end `)#` {left:?} {s:?}");
+                } else {
+                    while let Some(c) = iter.next() {
+                        if c == '|' {
+                            if let Some(cc) = iter.next_if_eq(&')') {
+                                if iter.next_if_eq(&'#').is_some() {
+                                    ranges.push(Range::new(start, left.len()));
+                                    continue 'outer;
+                                } else {
+                                    left.push(c);
+                                    left.push(cc);
+                                }
+                            } else {
+                                left.push(c);
+                            }
+                        } else {
+                            left.push(c);
+                        }
+                    }
+                    panic!("missing end `|)#` {left:?} {s:?}");
+                }
+            } else {
+                left.push(c);
+            }
+        } else {
+            left.push(c);
+        }
+    }
+    let primary = match primary {
+        Some(i) => i,
+        None => panic!("missing primary `#[|]#` {s:?}"),
+    };
+    let selection = Selection::new(ranges, primary);
+    (left, selection)
+}
```

Comment:
This seemed quite long, not sure if changing it to work with str will improve it or not but str we need to search it multiple times compared to this way.

Reply:
Just after a quick once over, this doesn't look *too* bad; it's not pretty, but it's not incomprehensible. It seems simple enough that I don't foresee a big maintenance problem, assuming we don't keep adding to the syntax, and I can't imagine how we would.

However, it is basically implementing a state machine parser by hand. I wonder if it would be overkill to pull in something like [`combine`](https://github.com/Marwes/combine) to offload the parsing. Taking a quick glance at the `Cargo.toml`, it appears that it only has one hard dependency on `memchr`, so it shouldn't be too bad as far as compile times. Just an idea though, I don't have a strong preference for this.

Reply:
Some early returns can make it more readable:

```rust
pub fn print(s: &str) -> (String, Selection) {
    let mut primary_idx = None;
    let mut ranges = SmallVec::new();
    let mut iter = s.chars().peekable();
    let mut left = String::with_capacity(s.len());

    'outer: while let Some(c) = iter.next() {
        let start = left.len();

        if c != '#' {
            left.push(c);
            continue;
        }

        let (is_primary, close_pair) = match iter.next() {
            Some('[') => (true, ']'),
            Some('(') => (false, ')'),
            Some(ch) => {
                left.push(ch);
                continue;
            }
            None => break,
        };

        if is_primary && primary_idx.is_some() {
            panic!("primary `#[` already appeared {left:?} {s:?}");
        }

        let head_at_beg = iter.next_if_eq(&'|').is_some();

        'in_range: while let Some(c) = iter.next() {
            if !(c == close_pair && iter.peek() == Some(&'#')) {
                left.push(c);
                continue 'in_range;
            }
            if !head_at_beg {
                let prev = left.pop().unwrap();
                if prev != '|' {
                    left.push(prev);
                    left.push(c);
                    continue 'in_range;
                }
            }
            iter.next(); // skip "#"

            if is_primary {
                primary_idx = Some(ranges.len());
            }
            let (anchor, head) = match head_at_beg {
                true => (left.len(), start),
                false => (start, left.len()),
            };
            ranges.push(Range::new(anchor, head));
            continue 'outer;
        }

        if head_at_beg {
            panic!("missing end `{close_pair}#` {left:?} {s:?}");
        } else {
            panic!("missing end `|{close_pair}#` {left:?} {s:?}");
        }
    }

    let primary = match primary_idx {
        Some(i) => i,
        None => panic!("missing primary `#[|]#` {s:?}"),
    };
    let selection = Selection::new(ranges, primary);
    (left, selection)
}
```

Reply:
Thanks, applied with some minor tweaks to remove inner loop label.